### PR TITLE
Adds template specialization for matrix_cl for double, var, and fvar

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This is a first draft of handling vars and fvars a little nicer in the `matrix_cl` class. I'm making this a PR on `bstatcomp` right now as this is just a PR for discussion.

 The goal here is that we can make things look a bit prettier by doing stuff like

```cpp
matrix_cl<double> A(eigen_double); // eigen_double is an eigen of doubles 
matrix_cl<var> B(eigen_var) // eigen_var is an eigen of vars
matrix_cl<var> C(10, 10) // var result matrix
```

`matrix_cl<var>` has two members `adj_` and `val_`, which we can access like in @andrjohns PR #1216
```cpp
C.val() = B.adj().triangular_transpose() * A;
```

I think that's pretty and should make the code for vars a bit easier to write. If we are clever enough this should make the standard API and OpenCL API close enough that we can go wide pretty quickly in adding more rev methods for the OpenCL backend.

The above is done through specializations for `double` and `var` of the matrix_cl class. This is brutal but effective imo and at least keeps things pretty simple.

`matrix_cl<double>` is our standard `matrix_cl` and all the kernel methods will only operate on the `matrix_cl<double>` specialization. The `matrix_cl<var>` specialization has members, `row_`, `col_`, `val_` and `adj_`, where `val_` and `adj_` are `matrix_cl<double>'s`.

With this change and specialization I'm starting to think that functions that use the `double` specialization should be in an `opencl/prim` folder. Alternatively, after this PR we can start moving the OpenCL methods to the respective `prim` and `rev` folders.  

In terms of `fvar`, so for I have this only working for `fvar<double>` and I should have `fvar<var>` and `fvar<fvar<var/double>>` done sometime next week. Same as Andfew's PR, the underlying `d_` and `val_` give back a `matrix_cl<T>` of the `fvar<T>`.

## Tests

Currently this only works with 

```
 ./runTests.py -j2 ./test/unit/math/opencl/add_test.cpp ./test/unit/math/opencl/opencl_matrix_test
```

I didn't want to do anymore tests until we decided this design is the right way to handle this.

I'm taggin Sean, Bob and, Andrew to skim this over. I wouldn't think of this as a full review but as a "does it seem like this is going in the right direction." since doing this all out is going to require me to change pretty much every file that uses a `matrix_cl`. In particular check out the matrix_cl.hpp file. If there's a more elegant way of doing this I'd be interested to hear about it.

### CODE REVIEW

**Comment by SteveBronder** [comment]
Ah, Bob and Andrew are not members in the fork so can't directly request a review, just pinging them

@bob-carpenter
@andrjohns

If yinz have time would be much appreciated! Mostly looking for, "The way you're doing this seems to make sense" or "The way you're doing this seems to be bad"

**Comment by rok-cesnovar** [comment]
I like this approach. It seems that it will simply the dealing with `Eigen<var>` and cant see that there is anything wrong with it. 

What is the plan here, to try and merge this before caching or the other way around? I would suggest to just open an "official" PR once you feel that its cleaned up enough.

**Comment by SteveBronder** [comment]
> I like this approach. It seems that it will simply the dealing with Eigen<var> and cant see that there is anything wrong with it.

Ty!

> What is the plan here, to try and merge this before caching or the other way around? I would suggest to just open an "official" PR once you feel that its cleaned up enough.

Tl;dr:
1. Andrew's PR
2. This
3. Cache

Unless the cache is v important for the glm. If it is then I can keep fixing the current cache branch and then update the glm branch. Otherwise it may be good to get the glm built off of develop

------

So as I've been cleaning up the cache PR I feel like it's going to be easy to goof when we start talking about vars and fvars. 

We only want to use the cache for Eigen matrices of doubles and I'd like that to be automated. If you look in cholesky theres a moment where we make an eigen matrix of doubles from the value of an eigen of vars. We should just make a matrix_cl<var>, transfer the eigen of vars over, then call `cholesky_decompose(A.val())`. Rn we would have to have a boolean the user sets in the constructor for matric_cl for eigen doubles that turns off (or on) the cache. I really dont like having to remember rules like that. 

With this pr I think it will be easier to setup the rule that constructing from an Eigen matrice of doubles sets the cache.

**Comment by rok-cesnovar** [comment]
Thanks for laying out the plan. GLMs do need the caching, but PRs for those will be open after caching, so no worries there. 

For the Stan GPU paper I am working on a branch that will have the current state of caching in so we can include GLMs. Will ping you if I need any help there.
